cmd/dra-example-kubeletplugin/driver.go

```diff
@@ -95,15 +95,28 @@ func (d *driver) Shutdown() error {
 	})
 }
 
-func (d *driver) NodePrepareResource(ctx context.Context, req *drapbv1.NodePrepareResourceRequest) (*drapbv1.NodePrepareResourceResponse, error) {
-	klog.Infof("NodePrepareResource is called: request: %+v", req)
+func (d *driver) NodePrepareResources(ctx context.Context, req *drapbv1.NodePrepareResourcesRequest) (*drapbv1.NodePrepareResourcesResponse, error) {
 
+	klog.Infof("NodePrepareResource is called: number of claims: %d", len(req.Claims))
+	preparedResources := &drapbv1.NodePrepareResourcesResponse{Claims: map[string]*drapbv1.NodePrepareResourceResponse{}}
+
+	// In production version some common operations of d.nodeUnprepareResources
+	// should be done outside of the loop, for instance updating the CR could
+	// be done once after all HW was prepared.
+	for _, claim := range req.Claims {
+		preparedResources.Claims[claim.Uid] = d.nodePrepareResources(ctx, claim)
+	}
+
+	return preparedResources, nil
+}
+
+func (d *driver) nodePrepareResources(ctx context.Context, claim *drapbv1.Claim) *drapbv1.NodePrepareResourceResponse {
 	var err error
 	var prepared []string
 	err = retry.RetryOnConflict(retry.DefaultRetry, func() error {
-		prepared, err = d.Prepare(req.ClaimUid)
+		prepared, err = d.prepare(claim.Uid)
 		if err != nil {
-			return fmt.Errorf("error allocating devices for claim '%v': %v", req.ClaimUid, err)
+			return fmt.Errorf("error allocating devices for claim '%v': %v", claim.Uid, err)
```

Comment:
Does this compile? The function returns `*drapbv1.NodePrepareResourceResponse`, not `error`.

I find it odd that `nodePrepareResources` takes a `drapbv1.Claim` but returns a `drapbv1.NodePrepareResourceResponse` - that's at two different levels of abstraction. I think I'd prefer returning CDI devices and an error here.

Reply:
This is an anonymous function that is called by RetryonConflict on line 103/116, its return value is indeed an error, which is checked on line 124 / 138.

Reply:
About the type of argument and return value - in main NodePrepareResources function we're looping through NodePrepareResourcesRequest.Claims[] to get respective entry for NodePrepareResourcesResponse.Claims[], the types used there happened to be named such way, although in here it might look odd - it looks completely find where this function is called.

Reply:
Okay.

Can you rebase now that #24 is merged?

Reply:
Rebase is not needed, this PR was built on top of #24

Reply:
The PR is currently shown as adding three commits, which is misleading. After a rebase, only one will be left.